DOC-757 | How-to on using CanvOS for building Edge native artifacts (#1318)
spectrocloud/librarium: content/docs/04-clusters/03-edge/04-edgeforge-workflow/03.5-palette-canvos.md

      tooltipPlacement: "rightTop",
    },
    {
      x: 190,
      y: 224,
      label: 2,
      description: "Instructs the installer to turn the host machine off once the installation is complete.",
    },
    {
      x: 190,
      y: 300,
      label: 3,
      description: "Sets the login credentials for Edge hosts. The login credentials will allow you to SSH log into the edge host for debugging purposes.",
      tooltipPlacement: "rightTop",
    },
  ]}
>

```shell
cat <<'EOF' > user-data
#cloud-config
stylus:
  site:
    edgeHostToken: aUAxxxxxxxxx0ChYCrO
install:
  poweroff: true
users:
  - name: kairos
    passwd: kairos
EOF
```

</PointsOfInterest>

## Build Edge Native Artifacts
Issue the following command to build the artifacts. 
<br />

```shell
./earthly.sh +build-all-images --PE_VERSION=$(git describe --abbrev=0 --tags)
```

```bash coloredLines=2-2|#006622
# Output condensed for readability
===================== Earthly Build SUCCESS ===================== 
Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev.
```
This command will take up to 15-20 minutes to finish.
<br />

> # Validate

Reviewer: This needs to be in numbered steps.

ttl.sh/ubuntu     k3s-1.25.2-v3.3.3-demo    b3c4956ccc0a    6 minutes ago   2.49GB
ttl.sh/ubuntu     k3s-1.24.6-v3.3.3-demo    fe1486da25df    6 minutes ago   2.49GB
earthly/earthly   v0.7.4                    d771cc8edc38     2 weeks ago    333MB
```

Suppose you want to use these provider images in your cluster profile. In that case, you can push them to the *ttl.sh* or any other image registry and add the following custom content within the `options` attribute of the **OS layer** of your cluster profile.    
<br />

```yaml
  system.uri: "{{ .spectro.pack.edge-native-byoi.options.system.registry }}/{{ .spectro.pack.edge-native-byoi.options.system.repo }}:{{ .spectro.pack.edge-native-byoi.options.system.k8sDistribution }}-{{ .spectro.system.kubernetes.version }}-{{ .spectro.pack.edge-native-byoi.options.system.peVersion }}-{{ .spectro.pack.edge-native-byoi.options.system.customTag }}"
  system.registry: ttl.sh
  system.repo: ubuntu
  system.k8sDistribution: k3s
  system.osName: ubuntu
  system.peVersion: v3.3.3
  system.customTag: demo
  system.osVersion: 22
```
You must verify and change these attributes' values, as applicable to you, before using them in your cluster profile. 

> <br />

Reviewer: We can link to the [Model Edge Native Cluster Profile](https://docs.spectrocloud.com/clusters/edge/site-deployment/model-profile) guide. The challenge with this validation is that the user would have to deploy an Edge host, deploy an Edge cluster to validate this. 

Instead, let's focus on ensuring the artifact is created and available in the registry.

<WarningBox>

If you are using *ttl.sh* image registry, your images will be available for a maximum of 24 hours.

</WarningBox>

The screenshot below shows custom content added to the OS layer of a cluster profile.
<br />

![Screenshot of k3s OS layer in a cluster profile](/tutorials/palette-canvos/clusters_edge_palette-canvos_edit_profile.png)

> # Cleanup

Reviewer: This is not part of the how-to layout.

  ├── Earthfile     # Series of commands for creating target images
  ├── earthly.sh    # Script to invoke the Earthfile, and generate target images
  └── user-data.template  # A sample user-data file
  ```
  <br />

5. This step is optional. In this step, you can edit the  **.arg** file to define arguments, such as, custom image tag for provider OS images, image registry, image repository, OS distribution, Kubernetes distribution, and ISO file name. 

  Skip this step if you use the following default values. 

  |**Argument**|**Description**|**Default Value**| **Other Possible Values** |
  |---|---|---|---|
  |`CUSTOM_TAG`|Tag for provider image|demo|Lowercase alphanumeric string without spaces|
  |`IMAGE_REGISTRY`|Image registry name|ttl.sh|Your image registry URL, without http or https <br /> Example: hub.docker.com|
  |`OS_DISTRIBUTION`|OS Distribution |ubuntu | opensuse-leap|
  |`IMAGE_REPO`|Image repository name|`$OS_DISTRIBUTION`|Your image repository name|
  |`OS_VERSION`|OS Version, only applies to Ubuntu |22| 20|
  |`K8S_DISTRIBUTION`|Kubernetes Distribution |k3s| rke2, kubeadm |
  |`ISO_NAME`|Name of the Installer ISO|palette-edge-installer|String without spaces|

>   Using the default value for all arguments, the provider image name will be similar to `ttl.sh/ubuntu:k3s-1.25.2-v3.4.1-demo`.

Reviewer: Don't use the defaults. Change it up so the users learn how to play with these values and to provide greater value than the basic section.

  To customize any of the arguments, edit its value in the **.arg** file. For example, if you want to change the image tag, edit the value for the `CUSTOM_TAG` variable. Ensure that it is of lowercase alphanumeric characters. 
  <br />

> 6. This step is also optional. In this step, you can modify the existing Dockerfile to suit your needs. For example, here is a sample Dockerfile to install [WireGuard](https://www.wireguard.com/install/) on the Edge host. 

Reviewer: Tell the user to open up the Dockerfile in an editor and to added the value in between the lines the CanvOS dockerfile instructs the user to update. We don't want the user to touch the lines above the ##### section. Let's be clear about that here.

  <br />

  ```Dockerfile
  ARG BASE
  FROM $BASE
  # Install WireGuard and required dependencies
  RUN apt-get update && apt-get install -y wireguard

  # Install Netplan
  RUN apt-get -y install netplan.io
  ```
  You can install more tools and dependencies and configure the host as needed.
  <br />

7. Create a registration token in Palette using the instructions available in the [Create Registration Token](/clusters/edge/site-deployment/site-installation/create-registration-token) guide to allow Edge hosts to register themselves with Palette automatically. 

>   Palette 3.4 onwards, a registration token created by the tenant admin, is now *required* for pairing an Edge host with Palette. Palette offers three registration methods: auto, manual, and QR code. This guide uses the auto-registration method. For more details, refer to the [Register Edge Host](https://docs.spectrocloud.com/clusters/edge/site-deployment/site-installation/edge-host-registration) documentation.

Reviewer: Make this the same as the feedback I provided in the basic section.

  <br />

  <PointsOfInterest
    points={[
      {
        x: 370,
        y: 185,
        label: 1,
        description: "Stores the registration token and lets the agent use the auto-registration functionality and authenticate with the provided token.",
        tooltipPlacement: "rightTop",
      },
      {
        x: 190,
        y: 262,
        label: 2,
        description: "Instructs the installer to turn the host machine off once the installation is complete.",
      },
      {
        x: 190,
        y: 340,
        label: 3,
        description: "Sets the login credentials for Edge hosts. The login credentials will allow you to SSH log into the edge host for debugging purposes.",
        tooltipPlacement: "rightTop",
      },
    ]}
  >

  ```shell
  cat <<'EOF' > user-data
  #cloud-config
  stylus:
    site:
      edgeHostToken: aUAxxxxxxxxx0ChYCrO
  install:
    poweroff: true
  users:
    - name: kairos
      passwd: kairos
  EOF
  ```

> 

Reviewer: Same feedback as basic section.

  </PointsOfInterest>
  <br />

9. Issue the following command to execute the **earthly.sh** file to build the Edge artifacts. The `--PE_VERSION` option in the command below signifies the Palette Edge version to use. Spectro Cloud generates a git tag for each new Palette release. The current example uses the **v3.3.3** git tag.
  <br />

  ```shell
  ./earthly.sh +build-all-images --PE_VERSION=$(git describe --abbrev=0 --tags)
  ```

  ```bash coloredLines=2-2|#666666
  # Output condensed for readability
  ===================== Earthly Build SUCCESS ===================== 
  Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev.
  ```

>   This command will take up to 15-20 minutes to finish.

Reviewer: Same feedback as basic section.

  ttl.sh/ubuntu     k3s-1.24.6-v3.3.3-palette-learn    fe1486da25df   6 minutes ago   2.49GB
  earthly/earthly   v0.7.4                             d771cc8edc38   2 weeks ago     333MB
  ```

</Tabs.TabPane>

<Tabs.TabPane tab="Advanced" key="advanced_create_artifacts">

1. Check out the [CanvOS](https://github.com/spectrocloud/CanvOS.git) GitHub repository containing the starter code. 
  <br />

  ```bash
  git clone https://github.com/spectrocloud/CanvOS.git
  ```
  

2. Change to the **CanvOS/** directory. 
  <br />

  ```bash
  cd CanvOS
  ```

3. The files relevant to the current guide are in the **v3.3.3** tag. Check out the desired tag. You can check out a more recent and stable version if available.
  <br />

  ```shell
  git checkout v3.3.3
  ```

4. Review the relevant files.  
  <br />

  ```bash
  .
  ├── .arg          # Defines mutiple variables
  ├── Dockerfile    # Bakes the arguments to use in the image
  ├── Earthfile     # Series of commands for creating target images
  ├── earthly.sh    # Script to invoke the Earthfile, and generate target images
  └── user-data.template  # A sample user-data file
  ```
  <br />

> 5. Edit the **.arg** file to customize the image tag for provider OS images, image registry, image repository, and OS distribution. 

Reviewer: Step five is way to large. Multiple steps are happening below. Break those out to its own number.

  |`OS_DISTRIBUTION`|OS Distribution |ubuntu | opensuse-leap|
  |`IMAGE_REPO`|Image repository name|`$OS_DISTRIBUTION`|Your image repository name|
  |`OS_VERSION`|OS Version, only applies to Ubuntu |22| 20|
  |`K8S_DISTRIBUTION`|Kubernetes Distribution |k3s| rke2, kubeadm |
  |`ISO_NAME`|Name of the Installer ISO|palette-edge-installer|String without spaces|

  Open up the **.arg** file in an editor and update the following variables. 
  <br />

    - Update the value for the `CUSTOM_TAG` variable to string made up of lowercase alphanumeric characters. This example uses `palette-learn` value. 
    <br />

    ```bash
    CUSTOM_TAG=palette-learn
    ```

    - Update the value for the `IMAGE_REGISTRY` to point it to the Docker Hub host. If you are using any other image registry, the hostname must comply with standard DNS rules, but may not contain underscores.
    <br />

    ```bash
    IMAGE_REGISTRY=registry.hub.docker.com
    ```

    - Update the value for the `OS_DISTRIBUTION` to OpenSuse Leap. This workflow supports OpenSuse Leap version 15.4.
    <br />

    ```bash
    OS_DISTRIBUTION=opensuse-leap
    ```

>     - Create a repository in the Docker Hub named `opensuse-leap` and set the repository's default visibility to public. Refer to [Create a repository](https://docs.docker.com/docker-hub/repos/create/#create-a-repository) instructions for creating a repository in the Docker Hub. The following declaration will default the repository name to be same as the OS distribtion. 

Reviewer: This should be a prereq. So the interesting thing is that we now have different requirements. So what we can do is move the Prerequisites section inside the <Tabs> element and assign each one their own Prerequisites.

<br />

* **Edge installer ISO image** - This bootable ISO image installs the necessary dependencies and configurations on a bare host machine. During installation, the host machine will boot from the Edge installer ISO, partition the disk, copy the image content to the disk, install the Palette Edge host agent and metadata, and perform several configuration steps. These configuration steps include registering the host with Palette, setting user privileges, and configuring network or security settings. 
<br />

* **Provider OS image** - This is a [Kairos](https://kairos.io/)-based image. This image installs an immutable Operating System (OS) and software dependencies compatible with a specific Kubernetes version at runtime, i.e., during the cluster deployment.  

In this guide, you will use Spectro Cloud's utility, [CanvOS](https://github.com/spectrocloud/CanvOS), to create an Edge installer ISO image and two provider OS images - one compatible with K3s v1.24.6 and another compatible with K3s v1.25.2. It builds two provider OS images, so you can use either of them that matches the Kubernetes version in your cluster profile.

> The diagram below shows the high-level steps to build the artifacts and push to an image registry.

Reviewer: The diagram is great but step 6 is what I worry could cause confusion. That's not done in this how-to, so I think it's best to remove it. Also, a USB is not the only way to load an Edge host, you've seen that with VMware.

  <br/>

  ```bash
  uname -m
  ```

* Minimum hardware configuration of the Linux machine:
  - 4 CPU
  - 8 GB memory
  - 50 GB storage
 

* [Git](https://cli.github.com/manual/installation). You can ensure git installation by issuing the `git --version` command.

* [Docker Engine](https://docs.docker.com/engine/install/) version 18.09.x or later. You can use the `docker --version` command to view the existing Docker version. You should have root-level or `sudo` privileges on your Linux machine to create privileged containers.   
<br />

  <WarningBox>

  Issue the following command to check if you can create privileged containers on your Linux machine.
  <br />

  ```bash
  docker run --privileged hello-world
  ```
  If the command above fails, create a `docker` user group and add the current user to it using the instructions available [here](https://docs.docker.com/engine/install/linux-postinstall/). Alternatively, if you do not want to create a user group, ensure you can issue Docker commands with `sudo` privileges on your Linux machine.

>   </WarningBox>

Reviewer: We don't need this, we just need to tell them to use `sudo`. Keep the bullet you have but let's remove this warning.

* Palette registration token for pairing Edge hosts with Palette. You will need tenant admin access to Palette to generate a new registration token. For detailed instructions, refer to the [Create Registration Token](/clusters/edge/site-deployment/site-installation/create-registration-token) guide.

# Instructions

Use the following instructions on your Linux machine to create all the required Edge artifacts quickly. 
<br />

1. Check out the [CanvOS](https://github.com/spectrocloud/CanvOS.git) GitHub repository containing the starter code. 
  <br />

  ```bash
  git clone https://github.com/spectrocloud/CanvOS.git
  ```

2. Change to the **CanvOS/** directory. 
  <br />

  ```bash
  cd CanvOS
  ```

3. View the available [git tags](https://github.com/spectrocloud/CanvOS/tags).
<br />

  ```bash
  git tags
  ```

4. Check out the repository tag that matches your Palette product version. Your Palette product version is displayed in the top left corner, as shown in the screenshot below. 

  ![Screenshot displaying the Palette product version.](/tutorials/palette-canvos/clusters_edge_palette-canvos_palette-version.png)

>   It is essential to check out a matching git tag so the generated Edge installer will be compatible with your Palette product version. This guide uses **v3.4.3** tag as an example. 

Reviewer: The first sentence is not written in an active voice. Address the user, speak to them directly. Starting a sentence with "is is essential" automatically puts a distance between the user and us.  If you're giving a recommendation, then say "We recommend"... otherwise it feels like we are trying to put space between what we say and what the user does.

  <br />

  ```shell
  git checkout v3.4.3
  ```

5. Review the files relevant for this guide. 
    - **.arg** - Defines arguments to use during the build process. 
    - **Dockerfile** - Bakes the arguments to use in the image.
    - **Earthfile** - Contains a series of commands to create target artifacts.
    - **earthly.sh** - Script to invoke the Earthfile, and generate target artifacts.
    - **user-data.template** - A sample user-data file
  <br />

> 6. Open the **.arg** file in an editor to change one of the arguments. 

Reviewer: This changed, review the new `.arg.template`

<br />

  ```bash
  vi .arg
  ```
  <br />

7. In the **.arg** file, change the value of the `CUSTOM_TAG` variable to any lowercase and alphanumeric string. This variable defines the custom image tag for the provider OS images. This guide uses the following custom tag as an example. 
<br /> 

  ```bash
  CUSTOM_TAG=palette-learn
  ```
  Save the file, and exit the editor. 

>   <br />

Reviewer: We need to change this. The feedback we got is that it felt weird jumping from the file editor to shell.  We decided to stick with doing all actions in the shell, so updating this value needs to happen without opening the file.

        tooltipPlacement: "rightTop",
      },
      {
        x: 190,
        y: 262,
        label: 2,
        description: "Instructs the installer to turn the host machine off once the installation is complete.",
      },
      {
        x: 190,
        y: 340,
        label: 3,
        description: "Sets the login credentials for Edge hosts. The login credentials will allow you to SSH log into the edge host for debugging purposes.",
        tooltipPlacement: "rightTop",
      },
    ]}
  >

  ```shell
  cat << EOF > user-data
  #cloud-config
  stylus:
    site:
      edgeHostToken: $token
  install:
    poweroff: true
  users:
    - name: kairos
      passwd: kairos
  EOF
  ```

  </PointsOfInterest>
  <br />

10. CanvOS utility internally uses [Earthly](https://earthly.dev/) to build the target artifacts. Issue the following command to invoke the Earthly build. The `--PE_VERSION` option in the command below signifies the Palette Edge version to use, which refers to the git tag you checked out in the previous step. 
<br />
  <WarningBox>

  Ensure to issue the command below with root-level or `sudo` privileges.

  </WarningBox>

  ```shell

>   ./earthly.sh +build-all-images --PE_VERSION=$(git describe --abbrev=0 --tags)

Reviewer: ```suggestion
  sudo ./earthly.sh +build-all-images --PE_VERSION=$(git describe --abbrev=0 --tags)
```
Let's just add the `sudo` and remove the warning box. Remember, people don't read, they will copy and paste.

  ```

  ```bash coloredLines=2-2
  # Output condensed for readability
  ===================== Earthly Build SUCCESS ===================== 
  Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev.
  ```
  This command may take up to 15-20 minutes to finish depending on the resources of the host machine.
  <br />

> 11. List the Docker images to verify the two provider OS images - one compatible with K3s v1.24.6 and another with K3s v1.25.2.

Reviewer: ```suggestion
11. List the Docker images to review the provider OS images created. By default, provider images for all the Edge-supported Kubernetes versions are created. You can identify the provider images by reviewing the image tag value you used in the  **.arg** file's `CUSTOM_TAG` variable. 
```

CanvOS creates all the images by default, so instead of mentioning two specifically, let's just teach the user to look for images with their tag. For example, you could do something like this

```
docker images --filter=reference='*palette-learn*'
```

  <br />

  ```bash
  docker push ttl.sh/ubuntu:k3s-1.25.2-v3.4.3-palette-learn
  docker push ttl.sh/ubuntu:k3s-1.24.6-v3.4.3-palette-learn
  ```
  <br />

  <WarningBox>

  As a reminder, [ttl.sh](https://ttl.sh/) is a short-lived image registry. If you do not use these provider OS images in your cluster profile within 24 hours of pushing to *ttl.sh*, they will expire and must be re-pushed. Refer to the Advanced workflow in the current guide to learn how to use another registry, such as Docker Hub, and tag the docker images accordingly.

  </WarningBox>
  <br />

13. After pushing the provider images to the image registry, add the following custom content within the `options` attribute of your cluster profile's **OS layer** so that the cluster profile can pull the provider OS image from the image registry. 

>   If you used customized arguments and Dockerfile to build the Edge artifacts, change these attributes' values, as applicable to you, before using them in your cluster profile.

Reviewer: We should mention that the OS pack should be the BYOOS pack.  We should also mention that the `system.XXXX` values come from what is used in the `.arg` file.

  <br />

  ```yaml
    system.uri: "{{ .spectro.pack.edge-native-byoi.options.system.registry }}/{{ .spectro.pack.edge-native-byoi.options.system.repo }}:{{ .spectro.pack.edge-native-byoi.options.system.k8sDistribution }}-{{ .spectro.system.kubernetes.version }}-{{ .spectro.pack.edge-native-byoi.options.system.peVersion }}-{{ .spectro.pack.edge-native-byoi.options.system.customTag }}"
    system.registry: ttl.sh
    system.repo: ubuntu
    system.k8sDistribution: k3s
    system.osName: ubuntu
    system.peVersion: v3.4.3
    system.customTag: demo
    system.osVersion: 22
  ```

>   The screenshot below shows custom content added to a cluster profile's OS layer.

Reviewer: ```suggestion
The following screenshot illustrates how custom content has been incorporated into an OS layer of a cluster profile using the [BYOOS](/integrations/byoos) pack.
```

Clarification of OS pack.

  ![Screenshot of a sample cluster profile's OS layer ](/tutorials/palette-canvos/clusters_edge_palette-canvos_edit_profile.png)
  <br />

>   After updating the OS layer, you must choose a Kubernetes version compatible with the provider OS image in your cluster profile. 

Reviewer: We can't stop here, have the user select a Kubernetes version and emphasize on using one of the versions created by the build process. And lastly, guide them in choosing a CNI but explain that they must choose a CNI that best fits their need.

# Build Edge Artifacts
Palette's Edge solution requires Edge hosts to be ready with the required dependencies and [user data](/clusters/edge/edge-configuration/installer-reference) configurations before deploying a Kubernetes cluster. An Edge host requires the following artifacts to prepare for successful cluster deployment:

<br />

* **Edge installer ISO image** - This bootable ISO image installs the necessary dependencies and configurations on a bare host machine. During installation, the host machine will boot from the Edge installer ISO, partition the disk, copy the image content to the disk, install the Palette Edge host agent and metadata, and perform several configuration steps. These configuration steps include registering the host with Palette, setting user privileges, and configuring network or security settings. 
<br />

* **Provider OS image** - This is a [Kairos](https://kairos.io/)-based image. This image installs an immutable Operating System (OS) and software dependencies compatible with a specific Kubernetes version at runtime, i.e., during the cluster deployment. A provider OS image is used in the OS layer while creating a cluster profile. 

> In this guide, you will use Spectro Cloud's utility, [CanvOS](https://github.com/spectrocloud/CanvOS/blob/main/README.md), to build an Edge installer ISO image and multiple provider OS images for all the Palette-supported Kubernetes versions. It builds multiple provider OS images, so you can use either one that matches the Kubernetes version in your cluster profile. This guide assumes that your Palette supports K3s v1.24.6 and K3s v1.25.2, as an example, for deploying Edge clusters.  

Reviewer: ```suggestion
In this guide, you will use Spectro Cloud's utility, [CanvOS](https://github.com/spectrocloud/CanvOS/blob/main/README.md), to build an Edge installer ISO image and multiple provider OS images for all the Palette-supported Kubernetes versions. The utility builds multiple provider OS images, so you can use either one that matches the desired Kubernetes version you want to use with your cluster profile.  
```

> The diagram below shows the high-level steps to build the artifacts and push the provider OS images to an image registry.

Reviewer: ```suggestion
The diagram below shows the high-level steps to building the Edge artifacts and pushing the provider OS images to an image registry.
```

20. Lastly, add the network layer to your cluster profile, and choose a Container Network Interface (CNI) pack that best fits your needs, such as Calico, Flannel, Cilium, or Custom CNI. 

  This step completes the core infrastructure layers in the cluster profile. Next, you can add add-on layers to your cluster profile. 
  <br />

## Validate
List the Edge installer ISO image and checksum by issuing the following command from the **CanvOS/** directory.
<br />

```shell
ls build/
```

```shell hideClipboard
# Output
palette-edge-installer.iso      
palette-edge-installer.iso.sha256
```

You can validate the ISO image by creating a bootable USB flash drive using any third-party software and attempting to flash a bare host machine. Most software that creates a bootable USB drive will validate the ISO image. Here, the flash process means installing the necessary tools and configurations on a  host machine.

</Tabs.TabPane>

</Tabs>

# Next Steps

> After building the Edge artifacts, the next step is to use the ISO image to prepare your Edge host. Check out the [Prepare Edge Host for Installation](/clusters/edge/site-deployment/stage) guide to learn more about the next steps.

Reviewer: ```suggestion
After building the Edge artifacts, the next step is to use the Edge Installer ISO image to prepare your Edge host. Check out the [Prepare Edge Host for Installation](/clusters/edge/site-deployment/stage) guide to learn more about the next steps.
```

  <br />

6. Issue the command below to save a custom image tag for the provider OS images in the `CUSTOM_TAG` variable. This guide uses `palette-learn` as an example. However, you can assign any lowercase and alphanumeric string to the `CUSTOM_TAG` variable. 
<br />

  ```bash
  export CUSTOM_TAG=palette-learn
  ```
  <br />

7. Issue the command below to create the **.arg** file with the custom tag. It uses the default values for the remaining variables. You can refer to the existing **.arg.template** file in the present working directory to know the variables you can customize.  
<br />

  ```bash
  cat << EOF > .arg
  CUSTOM_TAG=$CUSTOM_TAG
  IMAGE_REGISTRY=ttl.sh
  OS_DISTRIBUTION=ubuntu
  IMAGE_REPO=$OS_DISTRIBUTION
  OS_VERSION=22
  K8S_DISTRIBUTION=k3s
  ISO_NAME=palette-edge-installer
  EOF
  ```
  
  View the newly created file to ensure the customized variables are set correctly.
  <br />

  ```bash
  cat .arg
  ```
  <br />

8. Issue the command below to save your registration token to a local variable. Replace `[your_token_here]` with your actual registration token. 
<br />

  ```bash
  export token=[your_token_here]
  ```
  <br />
  

> 9. Issue the command below to create the **user-data** file that includes the tenant registration token. Also, you can click on the *Points of Interest* numbers below to learn more about the main attributes relevant to this example. 

Reviewer: ```suggestion
9. Use the following command to create the **user-data** file containing the tenant registration token. Also, you can click on the *Points of Interest* numbers below to learn more about the main attributes relevant to this example. 
```

20. Lastly, add the Network layer to your cluster profile, and choose a Container Network Interface (CNI) pack that best fits your needs, such as Calico, Flannel, Cilium, or Custom CNI. 

  This step completes the core infrastructure layers in the cluster profile. Next, you can add add-on layers to your cluster profile. 

# Validate

List the Edge installer ISO image and checksum by issuing the following command from the **CanvOS/** directory.
<br />

```shell
ls build/
```

```shell
# Output
palette-edge-installer.iso      
palette-edge-installer.iso.sha256
```

You can validate the ISO image by creating a bootable USB flash drive using any third-party software and attempting to flash a bare host machine. Most software that creates a bootable USB drive will also validate the ISO image. Here, the flash process means installing the necessary tools and configurations on a  host machine.

</Tabs.TabPane>

</Tabs>

# Next Steps

> After building the Edge artifacts, the next step is creating a bootable USB flash drive using the installer ISO and a cluster profile for your Edge hosts. You will reference the provider OS images you uploaded to the image registry in the cluster profile. Check out the [Model Edge Cluster Profile](/clusters/edge/site-deployment/model-profile) guide to learn more about the next steps.

Reviewer: ```suggestion
After building the Edge artifacts, the next step is to use the ISO image to prepare your Edge host. Check out the [Prepare Edge Host for Installation](/clusters/edge/site-deployment/stage) guide to learn more about the next steps.
```
